xbmc/video/videosync/VideoSyncIMX.cpp

```diff
@@ -0,0 +1,125 @@
+/*
+ *      Copyright (C) 2005-2014 Team XBMC
+ *      http://xbmc.org
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "system.h"
+
+#if defined(HAS_IMXVPU)
+
+#include "video/videosync/VideoSyncIMX.h"
+#include "guilib/GraphicContext.h"
+#include "windowing/WindowingFactory.h"
+#include "utils/TimeUtils.h"
+#include "utils/log.h"
+#include <linux/mxcfb.h>
+
+// TODO - once mxc_dcic driver integrated in kernel
+// mxc_dcic.h is exported as part of linux-libc-dev pkg
+// for now ease the compilation process by those
+// hard defined ioctls()
+//#include <linux/mxc_dcic.h>
+#define DCIC_IOC_CONFIG_DCIC    _IO('D', 12)
+#define DCIC_IOC_START_VSYNC    _IO('D', 15)
+#define DCIC_IOC_STOP_VSYNC     _IO('D', 16)
+// end TODO
+
+#include <sys/ioctl.h>
+#include "threads/Thread.h"
+
+bool CVideoSyncIMX::Setup(PUPDATECLOCK func)
+{
+  struct fb_var_screeninfo screen_info;
+
+  UpdateClock = func;
+  m_abort = false;
+
+  m_fddcic = open("/dev/mxc_dcic0", O_RDWR);
+  if (m_fddcic < 0)
+    return false;
+
+  int fb0 = open("/dev/fb0", O_RDONLY | O_NONBLOCK);
+  if (fb0 < 0)
+    return false;
+
+  bool bContinue = !ioctl(fb0, FBIOGET_VSCREENINFO, &screen_info);
+  if (bContinue) {
+    bContinue = !ioctl(m_fddcic, DCIC_IOC_CONFIG_DCIC, &screen_info.sync);
+
+    if (bContinue)
+      bContinue = !ioctl(m_fddcic, DCIC_IOC_START_VSYNC, 0);
+  }
+
+  close(fb0);
+  if (!bContinue)
+  {
+    if (m_fddcic > 0)
+      close(m_fddcic);
+    return false;
+  }
+
+  g_Windowing.Register(this);
+  CLog::Log(LOGDEBUG, "CVideoReferenceClock: setting up IMX");
+  return true;
+}
+
+void CVideoSyncIMX::Run(volatile bool& stop)
+{
+  unsigned long counter;
+  unsigned long last = 0;
+  /* This shouldn't be very busy and timing is important so increase priority */
+  CThread::GetCurrentThread()->SetPriority(CThread::GetCurrentThread()->GetPriority()+1);
```

Comment:
> btw: how can the SetPriority be called only like that SetPriority)() ? 

right. my mistake I did not look closely enough. The point is that no other class is supposed to change a thread's priority.

I have limited thread priorities for Linux to +/- 1 of application's priority. If this will change, this code will break.

If you want to raise prio, you should do this VideoReferenceClock which owns the thread.